cmd/process-agent/collector_api_test.go

```diff
@@ -19,56 +22,222 @@ import (
+
+		_, ok := reqBody.Body.(*process.CollectorContainer)
+		require.True(t, ok)
+	})
 }
 
-func TestSendContainerMessage(t *testing.T) {
+func TestSendProcMessage(t *testing.T) {
+	runCollectorTest(t, func(payloads chan checkPayload, cfg *config.AgentConfig, ep *mockEndpoint) {
+		m := &process.CollectorProc{
+			HostName: cfg.HostName,
+			GroupId:  1,
+			Containers: []*process.Container{
+				{Id: "1", Name: "foo"},
+			},
+		}
+
+		payloads <- checkPayload{
+			name:     checks.Process.Name(),
+			messages: []process.MessageBody{m},
+		}
+
+		req := <-ep.Requests
+
+		assert.Equal(t, "/api/v1/collector", req.uri)
+
+		assert.Equal(t, cfg.HostName, req.headers.Get(api.HostHeader))
+		assert.Equal(t, cfg.APIEndpoints[0].APIKey, req.headers.Get("DD-Api-Key"))
+		assert.Equal(t, "1", req.headers.Get(api.ContainerCountHeader))
+		assert.Equal(t, "1", req.headers.Get("X-DD-Agent-Attempts"))
+		assert.NotEmpty(t, req.headers.Get("X-DD-Agent-Timestamp"))
+
+		reqBody, err := process.DecodeMessage(req.body)
+		require.NoError(t, err)
+
+		_, ok := reqBody.Body.(*process.CollectorProc)
+		require.True(t, ok)
+	})
+}
+
+func TestSendProcMessageWithRetry(t *testing.T) {
+	runCollectorTest(t, func(payloads chan checkPayload, cfg *config.AgentConfig, ep *mockEndpoint) {
+		ep.ErrorCount = 1
+
+		m := &process.CollectorProc{
+			HostName: cfg.HostName,
+			GroupId:  1,
+			Containers: []*process.Container{
+				{Id: "1", Name: "foo"},
+			},
+		}
+
+		payloads <- checkPayload{
+			name:     checks.Process.Name(),
+			messages: []process.MessageBody{m},
+		}
+
+		requests := []request{
+			<-ep.Requests,
+			<-ep.Requests,
+		}
+
+		timestamps := make(map[string]struct{})
+		for _, req := range requests {
+			assert.Equal(t, cfg.HostName, req.headers.Get(api.HostHeader))
+			assert.Equal(t, cfg.APIEndpoints[0].APIKey, req.headers.Get("DD-Api-Key"))
+			assert.Equal(t, "1", req.headers.Get(api.ContainerCountHeader))
+			timestamps[req.headers.Get("X-DD-Agent-Timestamp")] = struct{}{}
+
+			reqBody, err := process.DecodeMessage(req.body)
+			require.NoError(t, err)
+
+			_, ok := reqBody.Body.(*process.CollectorProc)
+			require.True(t, ok)
+		}
+
+		assert.Len(t, timestamps, 1)
+		assert.Equal(t, "1", requests[0].headers.Get("X-DD-Agent-Attempts"))
+		assert.Equal(t, "2", requests[1].headers.Get("X-DD-Agent-Attempts"))
+	})
+}
+
+func TestRTProcMessageNotRetried(t *testing.T) {
+	runCollectorTest(t, func(payloads chan checkPayload, cfg *config.AgentConfig, ep *mockEndpoint) {
+		ep.ErrorCount = 1
+
+		m := &process.CollectorRealTime{
+			HostName: cfg.HostName,
+			GroupId:  1,
+		}
+
+		payloads <- checkPayload{
+			name:     checks.RTProcess.Name(),
+			messages: []process.MessageBody{m},
+		}
+
+		req := <-ep.Requests
+
+		reqBody, err := process.DecodeMessage(req.body)
+		require.NoError(t, err)
+
+		_, ok := reqBody.Body.(*process.CollectorRealTime)
+		require.True(t, ok)
+
+		assert.Equal(t, "1", req.headers.Get("X-DD-Agent-Attempts"))
+
+		select {
```

Comment:
could we have runCollectorTest read until `payloads` is closed and then flush to get rid of the need to sleep here?

Reply:
The sleep isn't ideal but I'm not aware a better way to test that something _doesn't_ happen, when accounting for an asynchronous i/o call like this.

We don't have access to the `payloads` channel here unless I'm misunderstanding what you're suggesting.